File: lib/factorix/blueprint.rb
# frozen_string_literal: true

require "base64"
require "json"
require "zlib"

module Factorix
  # Represents a Factorio blueprint
  #
  # A blueprint string has the format: version_byte + Base64(zlib(JSON))
  # Only version byte '0' is supported.
  class Blueprint
    # The only supported version byte
    SUPPORTED_VERSION = "0"
    private_constant :SUPPORTED_VERSION

    # @return [Hash] The blueprint data
    attr_reader :data

    # Decode a blueprint string into a Blueprint
    #
    # @param string [String] The blueprint string
    # @return [Blueprint]
    # @raise [UnsupportedBlueprintVersionError] if the version byte is not '0'
    # @raise [BlueprintFormatError] if the string is not a valid blueprint
    def self.decode(string)
      version = string[0]
      raise UnsupportedBlueprintVersionError, "Unsupported blueprint version: #{version.inspect}" unless version == SUPPORTED_VERSION

      compressed = Base64.strict_decode64(string[1..])
      json_string = Zlib::Inflate.inflate(compressed)
      new(JSON.parse(json_string))
    rescue ArgumentError => e
      raise BlueprintFormatError, "Invalid Base64 encoding: #{e.message}"
    rescue Zlib::Error => e
      raise BlueprintFormatError, "Invalid zlib data: #{e.message}"
    rescue JSON::ParserError => e
      raise BlueprintFormatError, "Invalid JSON: #{e.message}"
    end

    # @param data [Hash] The blueprint data
    def initialize(data)
      @data = data
    end

    # Encode this blueprint to a blueprint string
    #
    # @return [String]
    def encode
      json_string = JSON.generate(@data)
      compressed = Zlib::Deflate.deflate(json_string, Zlib::BEST_COMPRESSION)
      SUPPORTED_VERSION + Base64.strict_encode64(compressed)
    end

    # Serialize this blueprint to pretty-printed JSON
    #
    # @return [String]
    def to_json(*)
      JSON.pretty_generate(@data)
    end
  end
end

File: internal/blueprint/blueprint.go
// Package blueprint encodes and decodes Factorio blueprint strings
// (a version byte followed by Base64-encoded zlib-compressed JSON).
package blueprint

import (
	"bytes"
	"compress/zlib"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

// The only supported version byte.
const supportedVersion = '0'

var (
	ErrUnsupportedVersion = errors.New("unsupported blueprint version")
	ErrFormat             = errors.New("invalid blueprint")
)

// Blueprint holds the blueprint's JSON. The original text is kept as-is so
// key order survives a decode/encode round trip.
type Blueprint struct {
	json []byte
}

// Decode parses a blueprint string.
func Decode(s string) (*Blueprint, error) {
	if len(s) == 0 || s[0] != supportedVersion {
		version := "(empty)"
		if len(s) > 0 {
			version = string(s[0])
		}
		return nil, fmt.Errorf("%w: %s", ErrUnsupportedVersion, version)
	}

	compressed, err := base64.StdEncoding.DecodeString(s[1:])
	if err != nil {
		return nil, fmt.Errorf("%w: invalid Base64 encoding: %s", ErrFormat, err)
	}
	r, err := zlib.NewReader(bytes.NewReader(compressed))
	if err != nil {
		return nil, fmt.Errorf("%w: invalid zlib data: %s", ErrFormat, err)
	}
	defer r.Close()
	jsonData, err := io.ReadAll(r)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid zlib data: %s", ErrFormat, err)
	}
	return FromJSON(jsonData)
}

// FromJSON builds a Blueprint from JSON content.
func FromJSON(data []byte) (*Blueprint, error) {
	if !json.Valid(data) {
		return nil, fmt.Errorf("%w: invalid JSON", ErrFormat)
	}
	return &Blueprint{json: data}, nil
}

// Encode renders the blueprint string.
func (b *Blueprint) Encode() (string, error) {
	var compact bytes.Buffer
	if err := json.Compact(&compact, b.json); err != nil {
		return "", fmt.Errorf("%w: %s", ErrFormat, err)
	}

	var compressed bytes.Buffer
	w, err := zlib.NewWriterLevel(&compressed, zlib.BestCompression)
	if err != nil {
		return "", err
	}
	if _, err := w.Write(compact.Bytes()); err != nil {
		return "", err
	}
	if err := w.Close(); err != nil {
		return "", err
	}
	return string(supportedVersion) + base64.StdEncoding.EncodeToString(compressed.Bytes()), nil
}

// JSON renders the blueprint content as pretty-printed JSON, preserving the
// original key order.
func (b *Blueprint) JSON() ([]byte, error) {
	var pretty bytes.Buffer
	if err := json.Indent(&pretty, b.json, "", "  "); err != nil {
		return nil, err
	}
	return pretty.Bytes(), nil
}
